pkg/extensions/tbtc/tbtc.go

```diff
@@ -0,0 +1,292 @@
+	actBackoffFn backoffFn,
+	timeout time.Duration,
+) error {
+	monitoringSubscription, err := t.monitorAndAct(
+		ctx,
+		"retrieve pubkey",
+		func(handler depositEventHandler) (subscription.EventSubscription, error) {
+			return t.chain.OnDepositCreated(handler)
+		},
+		func(handler depositEventHandler) (subscription.EventSubscription, error) {
+			return t.chain.OnDepositRegisteredPubkey(handler)
+		},
+		t.watchKeepClosed,
+		func(deposit string) error {
+			return t.chain.RetrieveSignerPubkey(deposit)
+		},
+		actBackoffFn,
+		timeout,
+	)
+	if err != nil {
+		return err
+	}
+
+	go func() {
+		<-ctx.Done()
+		monitoringSubscription.Unsubscribe()
+		logger.Infof("retrieve pubkey monitoring disabled")
+	}()
+
+	logger.Infof("retrieve pubkey monitoring initialized")
+
+	return nil
+}
+
+// TODO:
+//  1. Filter incoming events by operator interest.
+//  2. Incoming events deduplication.
+//  3. Resume monitoring after client restart.
+func (t *tbtc) monitorAndAct(
+	ctx context.Context,
+	monitoringName string,
+	monitoringStartFn watchDepositEventFn,
+	monitoringStopFn watchDepositEventFn,
+	keepClosedFn watchKeepClosedFn,
+	actFn submitDepositTxFn,
+	actBackoffFn backoffFn,
+	timeout time.Duration,
+) (subscription.EventSubscription, error) {
+	handleStartEvent := func(deposit string) {
+		logger.Infof(
+			"starting [%v] monitoring for deposit [%v]",
+			monitoringName,
+			deposit,
+		)
+
+		stopEventChan := make(chan struct{})
+
+		stopEventSubscription, err := monitoringStopFn(
+			func(stopEventDeposit string) {
+				if deposit == stopEventDeposit {
+					stopEventChan <- struct{}{}
+				}
+			},
+		)
+		if err != nil {
+			logger.Errorf(
+				"could not setup stop event handler for [%v] "+
+					"monitoring for deposit [%v]: [%v]",
+				monitoringName,
+				deposit,
+				err,
+			)
+			return
+		}
+		defer stopEventSubscription.Unsubscribe()
+
+		keepClosedChan, keepClosedUnsubscribe, err := keepClosedFn(deposit)
+		if err != nil {
+			logger.Errorf(
+				"could not setup keep closed handler for [%v] "+
+					"monitoring for deposit [%v]: [%v]",
+				monitoringName,
+				deposit,
+				err,
+			)
+			return
+		}
+		defer keepClosedUnsubscribe()
+
+		timeoutChan := time.After(timeout)
+
+		actionAttempt := 1
+
+	monitoring:
+		for {
+			select {
+			case <-ctx.Done():
+				logger.Infof(
+					"context is done for [%v] "+
+						"monitoring for deposit [%v]",
+					monitoringName,
+					deposit,
+				)
+				break monitoring
+			case <-stopEventChan:
+				logger.Infof(
+					"stop event occurred for [%v] "+
+						"monitoring for deposit [%v]",
+					monitoringName,
+					deposit,
+				)
+				break monitoring
+			case <-keepClosedChan:
+				logger.Infof(
+					"keep closed event occurred for [%v] "+
+						"monitoring for deposit [%v]",
+					monitoringName,
+					deposit,
+				)
+				break monitoring
+			case <-timeoutChan:
+				logger.Infof(
+					"timeout occurred for [%v] monitoring for "+
+						"deposit [%v]; performing action",
```

Comment:
Nitpick: When I see `timeout occurred` in logs, I always assume something went wrong while in this case it's a happy path. I think we should avoid logging *timeout* here. Maybe:
```
`[%v] not performed by depositor in the expected time frame for deposit [%v], acting on their behalf`
```

Reply:
I will change the `not performed by depositor` to `not performed by externally` in case of other actions this will not be always a depositor I think.

Reply:
Or let's just have `not performed` to not suggest anything. It's good enough.